csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -721,6 +727,65 @@ public void CopyTo(Array array, int index)
             }
         }
 
+        /// <summary>
+        /// 
+        /// </summary>
+        public readonly struct Enumerable : IEnumerable<KeyValuePair<TKey, TValue>>
+        {
+            private readonly MapField<TKey, TValue> field;
+
+            internal Enumerable(MapField<TKey, TValue> field)
+            {
+                this.field = field;
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            /// <returns></returns>
+            public Enumerator GetEnumerator() => new Enumerator(field);
+
+            IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => GetEnumerator();
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private LinkedList<KeyValuePair<TKey, TValue>>.Enumerator enumerator;
+
+            internal Enumerator(MapField<TKey, TValue> field)
+            {
+                this.enumerator = field.list.GetEnumerator();
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            public KeyValuePair<TKey, TValue> Current => enumerator.Current;
+
+            object IEnumerator.Current => Current;
+
+            /// <summary>
+            /// 
+            /// </summary>
+            public void Dispose() => enumerator.Dispose();
+
+            /// <summary>
+            /// 
+            /// </summary>
+            /// <returns></returns>
+            public bool MoveNext() => enumerator.MoveNext();
+
+            /// <summary>
+            /// 
+            /// </summary>
+            void IEnumerator.Reset() => throw new NotSupportedException();
```

Comment:
Wondering why you don't keep the list around so you can implement reset?

Reply:
I wondered about that. My gut feeling is that Reset is basically used by *no-one*, and was an API mistake, so I didn't want to make anything even slightly more complex to support it.